File: apps/web/src/features/markers/weatherPin.test.ts
import { describe, it, expect } from 'vitest';
import {
  WEATHER_PIN_STALE_MS,
  WEATHER_PIN_ICON_PREFIX,
  weatherPinFetchDecision,
  weatherPinUiState,
  encodeWireIcon,
  decodeWireIcon,
  type WeatherSnapshot,
} from './weatherPin';

const snap = (over: Partial<WeatherSnapshot> = {}): WeatherSnapshot => ({
  temperatureC: 7,
  symbolCode: 'lightrain',
  windSpeedMs: 4,
  windFromDeg: 270,
  precipitationMm: 0.4,
  ...over,
});

/** When a weather pin opens/drops it decides whether to refresh from MET or
 *  render its cached forecast — the behaviour that keeps it live yet
 *  offline-safe. */
describe('weatherPinFetchDecision', () => {
  const HOUR = WEATHER_PIN_STALE_MS;

  it('never fetches offline — even with no cache, a fetch cannot succeed', () => {
    expect(weatherPinFetchDecision(null, false)).toBe('UseCache');
    expect(weatherPinFetchDecision(5 * HOUR, false)).toBe('UseCache');
  });

  it('fetches online when there is no cached forecast', () => {
    expect(weatherPinFetchDecision(null, true)).toBe('Fetch');
  });

  it('fetches online when the cache is older than an hour', () => {
    expect(weatherPinFetchDecision(HOUR + 1, true)).toBe('Fetch');
  });

  it('keeps the cache online when it is still fresh', () => {
    expect(weatherPinFetchDecision(HOUR - 1, true)).toBe('UseCache');
    expect(weatherPinFetchDecision(HOUR, true)).toBe('UseCache');
    expect(weatherPinFetchDecision(0, true)).toBe('UseCache');
  });

  it('honours a custom staleness window', () => {
    expect(weatherPinFetchDecision(10_000, true, 5_000)).toBe('Fetch');
    expect(weatherPinFetchDecision(4_000, true, 5_000)).toBe('UseCache');
  });
});

/** What the on-map glyph and its expanded card show, derived from the cached
 *  forecast — including how long ago it was updated and whether marine data is
 *  available. */
describe('weatherPinUiState', () => {
  const NOW = 1_000_000_000_000;

  it('is null when the pin has no cached forecast yet', () => {
    expect(weatherPinUiState({}, NOW)).toBeNull();
    expect(weatherPinUiState({ forecastFetchedAtEpochMs: NOW }, NOW)).toBeNull();
  });

  it('surfaces temperature, condition glyph code and the expanded air fields', () => {
    const s = weatherPinUiState(
      { forecast: snap(), forecastFetchedAtEpochMs: NOW },
      NOW,
    )!;
    expect(s.temperatureC).toBe(7);
    expect(s.symbolCode).toBe('lightrain');
    expect(s.expanded.windSpeedMs).toBe(4);
    expect(s.expanded.windFromDeg).toBe(270);
    expect(s.expanded.precipitationMm).toBe(0.4);
  });

  it('reports whole hours since the cache was written', () => {
    const threeHalfHoursAgo = NOW - 3.5 * WEATHER_PIN_STALE_MS;
    const s = weatherPinUiState({ forecast: snap(), forecastFetchedAtEpochMs: threeHalfHoursAgo }, NOW)!;
    expect(s.updatedHoursAgo).toBe(3);
  });

  it('has a null updated age when the cache carries no timestamp', () => {
    const s = weatherPinUiState({ forecast: snap() }, NOW)!;
    expect(s.updatedHoursAgo).toBeNull();
  });

  it('flags marine data only when wave/sea-temperature is present', () => {
    const inland = weatherPinUiState({ forecast: snap(), forecastFetchedAtEpochMs: NOW }, NOW)!;
    expect(inland.hasMarine).toBe(false);

    const atSea = weatherPinUiState(
      { forecast: snap({ waveHeightM: 1.2, waveFromDeg: 200, seaTemperatureC: 9 }), forecastFetchedAtEpochMs: NOW },
      NOW,
    )!;
    expect(atSea.hasMarine).toBe(true);
    expect(atSea.expanded.waveHeightM).toBe(1.2);
    expect(atSea.expanded.seaTemperatureC).toBe(9);
  });
});

/** The marker kind rides on the wire `icon` (the geo location has no kind
 *  field), so a weather pin must survive a sync round-trip instead of coming
 *  back as a plain marker — the latent data-loss fix. */
describe('wire-icon codec', () => {
  it('namespaces a weather pin, leaving standard markers untouched', () => {
    expect(encodeWireIcon('hiking', 'WeatherPin')).toBe(`${WEATHER_PIN_ICON_PREFIX}hiking`);
    expect(encodeWireIcon('hiking', 'Standard')).toBe('hiking');
  });

  it('decodes the prefix back to a weather pin and its activity icon', () => {
    expect(decodeWireIcon('weatherpin:kayaking')).toEqual({ activityIcon: 'kayaking', markerKind: 'WeatherPin' });
    expect(decodeWireIcon('camping')).toEqual({ activityIcon: 'camping', markerKind: 'Standard' });
  });

  it('defaults a blank icon to a standard mountain pin', () => {
    expect(decodeWireIcon('')).toEqual({ activityIcon: 'mountain', markerKind: 'Standard' });
  });

  it('round-trips both kinds through encode → decode', () => {
    for (const [icon, kind] of [
      ['hiking', 'Standard'],
      ['kayaking', 'WeatherPin'],
      ['fishing', 'WeatherPin'],
    ] as const) {
      expect(decodeWireIcon(encodeWireIcon(icon, kind))).toEqual({ activityIcon: icon, markerKind: kind });
    }
  });
});

File: apps/android/core/model/src/test/kotlin/com/sigmundgranaas/turbo/expressive/domain/WeatherPinTest.kt
package com.sigmundgranaas.turbo.expressive.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Behavior of the weather-pin seams: when a pin re-fetches vs. paints from cache, and what
 * its cached forecast renders as. These are the pure units the ViewModel and render layer
 * sit on top of — asserted here without a device.
 */
class WeatherPinTest {

    private val hour = 60L * 60L * 1000L

    // ── weatherPinFetchDecision: the "use cache vs refetch" reducer ──

    @Test
    fun `offline always keeps the cache even when stale`() {
        assertEquals(WeatherPinFetch.UseCache, weatherPinFetchDecision(cacheAgeMs = 6 * hour, online = false))
    }

    @Test
    fun `offline with no cache still does not fetch`() {
        assertEquals(WeatherPinFetch.UseCache, weatherPinFetchDecision(cacheAgeMs = null, online = false))
    }

    @Test
    fun `online with no cache yet fetches`() {
        assertEquals(WeatherPinFetch.Fetch, weatherPinFetchDecision(cacheAgeMs = null, online = true))
    }

    @Test
    fun `online with a fresh cache paints from cache`() {
        assertEquals(WeatherPinFetch.UseCache, weatherPinFetchDecision(cacheAgeMs = 10 * 60 * 1000L, online = true))
    }

    @Test
    fun `online with a stale cache refetches`() {
        assertEquals(WeatherPinFetch.Fetch, weatherPinFetchDecision(cacheAgeMs = 2 * hour, online = true))
    }

    @Test
    fun `exactly one hour old is still fresh (boundary is strictly greater-than)`() {
        assertEquals(WeatherPinFetch.UseCache, weatherPinFetchDecision(cacheAgeMs = hour, online = true))
    }

    // ── weatherPinUiState: what the cache renders as ──

    private val coastalSnapshot = WeatherSnapshot(
        temperatureC = -2.0,
        symbolCode = "cloudy",
        windSpeedMs = 7.5,
        windFromDeg = 315.0,
        precipitationMm = 1.2,
        waveHeightM = 1.8,
        waveFromDeg = 270.0,
        seaTemperatureC = 6.0,
    )

    @Test
    fun `standard view shows the cached temperature and condition glyph`() {
        val marker = weatherPin(coastalSnapshot, fetchedAt = 1_000L)
        val ui = weatherPinUiState(marker, nowMs = 1_000L)!!
        assertEquals(-2.0, ui.temperatureC!!, 1e-9)
        assertEquals("cloudy", ui.symbolCode)
    }

    @Test
    fun `expanded view surfaces wind, precipitation and marine fields`() {
        val ui = weatherPinUiState(weatherPin(coastalSnapshot, fetchedAt = 0L), nowMs = 0L)!!
        val e = ui.expanded
        assertEquals(7.5, e.windSpeedMs!!, 1e-9)
        assertEquals(315.0, e.windFromDeg!!, 1e-9)
        assertEquals(1.2, e.precipitationMm!!, 1e-9)
        assertEquals(1.8, e.waveHeightM!!, 1e-9)
        assertEquals(6.0, e.seaTemperatureC!!, 1e-9)
        assertTrue("coastal pin exposes marine data", e.hasMarine)
    }

    @Test
    fun `an inland cache reports no marine data`() {
        val inland = coastalSnapshot.copy(waveHeightM = null, seaTemperatureC = null, waveFromDeg = null)
        val ui = weatherPinUiState(weatherPin(inland, fetchedAt = 0L), nowMs = 0L)!!
        assertTrue("inland pin has no marine section", !ui.expanded.hasMarine)
    }

    @Test
    fun `updated-hours-ago reflects the fetch timestamp for the "updated Nh ago" cue`() {
        val ui = weatherPinUiState(weatherPin(coastalSnapshot, fetchedAt = 0L), nowMs = 3 * hour)!!
        assertEquals(3L, ui.updatedHoursAgo)
    }

    @Test
    fun `a pin with no cache yet renders nothing (still fetching)`() {
        val bare = Marker(id = "w", name = "Weather pin", kind = ActivityKindId.Viewpoint, position = LatLng(69.6, 18.9), markerKind = MarkerKind.WeatherPin)
        assertNull(weatherPinUiState(bare, nowMs = 5_000L))
    }

    @Test
    fun `snapshot projection keeps temp, wind, precip and marine from a live fetch`() {
        val conditions = Conditions(
            weather = WeatherNow(temperatureC = 4.0, windSpeedMs = 3.0, windFromDeg = 90.0, precipitationMm = 0.5, symbolCode = "rain"),
            avalanche = null,
            marine = MarineNow(waveHeightM = 2.1, waveFromDeg = 200.0, seaTemperatureC = 7.0),
        )
        val snap = WeatherSnapshot.from(conditions)
        assertEquals(4.0, snap.temperatureC!!, 1e-9)
        assertEquals(3.0, snap.windSpeedMs!!, 1e-9)
        assertEquals(0.5, snap.precipitationMm!!, 1e-9)
        assertEquals(2.1, snap.waveHeightM!!, 1e-9)
        assertEquals("rain", snap.symbolCode)
    }

    private fun weatherPin(snapshot: WeatherSnapshot, fetchedAt: Long) = Marker(
        id = "w-1",
        name = "Weather pin",
        kind = ActivityKindId.Viewpoint,
        position = LatLng(69.6, 18.9),
        markerKind = MarkerKind.WeatherPin,
        forecast = snapshot,
        forecastFetchedAtEpochMs = fetchedAt,
    )
}
